[Improvement] QgsGeos uses GEOSGeometry member of QgsGeometry (temporary solution)

ahuarte47 opened the pull request:
[QgsGeos](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeos.h#L27) class wraps the [QgsGeometryEngine](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeometryengine.h#L26) interface of a [QgsAbstractGeometryV2](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsabstractgeometryv2.h#L39) geometry. It is used massively for many spatial functions ([IsGeosValid](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeometry.cpp#L1564), [contains](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeometry.cpp#L821), crosses, intersects....) in [QgsGeometry](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeometry.h#L67).

But testing perfomance I see that all operations create in QgsGeos a new [GEOSGeometry](https://github.com/qgis/QGIS/blob/master/src/core/geometry/qgsgeos.cpp#L138) instance from the current QgsGeometry object.

It is a performance bottleneck (e.g. labeling), We can test....

```
QgsGeometry a = ...
QgsGeometry b = ...

for (int  i = 0; i < 100; i++)
{
  bool contains = a.contains( &b );
}
```

The current code creates 200 GEOSGeometry objects.
I am right ? Is there one reason for this behavior?

nyalldawson commented:
@ahuarte47 there's some related discussion here: 
https://lists.osgeo.org/pipermail/qgis-developer/2015-June/038283.html

ahuarte47 commented:
Thank @nyalldawson, sorry I did not see this discussion. Then this pull can fix it :-)

m-kuhn commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -1583,13 +1583,13 @@ bool QgsGeometry::isGeosEmpty() const
 QgsGeometry *QgsGeometry::unaryUnion( const QList<QgsGeometry*> &geometryList )
 {
-  QgsGeos geos( nullptr );
+  QgsGeos geos( (QgsAbstractGeometryV2*)( nullptr ) );
```
Is there something wrong with the old call?

m-kuhn commented:
There was a big discussion recently concerning geometry engine etc https://github.com/qgis/QGIS/pull/2660 

I am in favor of caching the geos geometry (and possibly others) in QgsGeometry. I would probably even directly return a QgsGeometry instead of an abstract geometry pointer from all methods that create a new geometry, given that they probably already have a geos geometry created while creating the new geometry.

QgsGeometry should most of the time not be passed as pointer but either const reference parameter or by value returned (like QString) since it's implicitly shared already.

ahuarte47 commented:
Thanks Mathias, I see it. Sorry, I precipitated myself doing this work, I did not search if there is already work done. I think to close my pull is better to avoid new noise. Do you agree?

m-kuhn commented:
I don't think anybody is working on it, so if you feel like tackling this, feel free :)

ahuarte47 commented:
I see the [work](https://github.com/qgis/QGIS/pull/2660#issuecomment-170910643), my pull is a temporary solution (I do not know if it's worth while QgsSimpleFeatureGeometryEngine... may not be implemented). Anyway, I will try to implement it if I find time.

ahuarte47 committed:
QgsGeos uses the GEOSGeometry member of QgsGeometry

QgsGeos references the GEOSGeometry pointer of QgsGeometry to avoid
repeated creations.
It is a temporary solution while QgsSimpleFeatureGeometryEngine and
related classes may not be implemented. See:
https://github.com/qgis/QGIS/pull/2660#issuecomment-170910643

ahuarte47 commented:
Hi @nyalldawson and @m-kuhn, I started the [work](https://github.com/qgis/QGIS/compare/master...ahuarte47:Issue_SimpleFeatureGeometryEngine).

Before I to continue, I would love if you may review it! I started [QgsSimpleFeatureGeometryEngine](https://github.com/qgis/QGIS/compare/master...ahuarte47:Issue_SimpleFeatureGeometryEngine#diff-1d8bad2726fa1f3b3af845454a6fcb85R21) and [QgsGeosGeometryEngine](https://github.com/qgis/QGIS/compare/master...ahuarte47:Issue_SimpleFeatureGeometryEngine#diff-400f602b09352bbe0c558e54e74ce036R23) classes with the intersection method. Also, I changed the geometry call to this new methods.

Thanks in advance!

ahuarte47 commented:
I close this pull, I will replace with this [work in progress](https://github.com/qgis/QGIS/compare/master...ahuarte47:Issue_SimpleFeatureGeometryEngine) that I will propose.

Thanks all!

ahuarte47 commented:
Hi all, is it possible to talk about this together in QGIS conference tomorrow ?
